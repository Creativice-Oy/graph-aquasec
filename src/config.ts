import {
  IntegrationExecutionContext,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';

/**
 * A type describing the configuration fields required to execute the
 * integration for a specific account in the data provider.
 *
 * When executing the integration in a development environment, these values may
 * be provided in a `.env` file with environment variables. For example:
 *
 * - `CLIENT_ID=123` becomes `instance.config.clientId = '123'`
 * - `CLIENT_SECRET=abc` becomes `instance.config.clientSecret = 'abc'`
 *
 * Environment variables are NOT used when the integration is executing in a
 * managed environment. For example, in JupiterOne, users configure
 * `instance.config` in a UI.
 */
export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  /**
   * NOTE: Aqua calls this an "API Key", but it is essentially an auto-generated
   * API key _name_. The value of this property is _visible_ from the Aqua
   * console. The actual sensitive token is the `apiSecret` described below.
   */
  apiKey: {
    type: 'string',
  },
  apiSecret: {
    type: 'string',
    mask: true,
  },
  accountId: {
    type: 'string',
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  apiKey: string;
  apiSecret: string;
  accountId: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationInstanceConfig>,
) {
  await Promise.resolve();
}
