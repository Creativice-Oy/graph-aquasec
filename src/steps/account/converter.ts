import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { AquaSecAccount } from '../../aquasec/types';

import { Entities } from '../constants';

export function createAccountEntity(data: AquaSecAccount): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: `aquasec_account:${data.id}`,
        _type: Entities.ACCOUNT._type,
        _class: Entities.ACCOUNT._class,
        id: data.id.toString(),
        name: data.id.toString(),
        owner: data.root_user,
        createdOn: parseTimePropertyValue(data.created),

        stripeId: data.stripe_id,
        trialEnd: parseTimePropertyValue(data.trial_end),
        currentPlan: data.current_plan,
        additionalCredits: data.additional_credits,
        numAllowedKeys: data.num_allowed_keys,
        suppressionsAdminOnly: data.suppressions_admin_only,
        globalSuppressionsAdminOnly: data.global_suppressions_admin_only,
        source: data.source,
        enforceOauth: data.enforce_oauth,
        usersAdminOnly: data.users_admin_only,
        integrationsAdminOnly: data.integrations_admin_only,
        disableDefaultEmails: data.disable_default_emails,
        blockDefaultGroup: data.block_default_group,
        presuppressPlugins: data.presuppress_plugins,
        associateTags: data.associate_tags,
        awsMarketplaceId: data.aws_marketplace_id,
        cspEnabled: data.csp_enabled,
        rootUser: data.root_user,
      },
    },
  });
}
