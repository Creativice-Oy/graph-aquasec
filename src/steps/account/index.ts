import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';
import { createApiClient } from '../../aquasec/jupiterone';
import { IntegrationConfig } from '../../config';
import { setAccountEntity } from '../../util/jobState';
import { Steps, Entities } from '../constants';
import { createAccountEntity } from './converter';

export async function fetchAccountDetails(
  context: IntegrationStepExecutionContext<IntegrationConfig>,
) {
  const { jobState, instance } = context;
  const client = createApiClient(context);
  const account = await client.getAccount(instance.config.accountId);

  await setAccountEntity(
    jobState,
    await jobState.addEntity(createAccountEntity(account)),
  );
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: Steps.ACCOUNT,
    name: 'Fetch Account Details',
    entities: [Entities.ACCOUNT],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
