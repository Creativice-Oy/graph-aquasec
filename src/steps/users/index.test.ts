import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../../config';
import { fetchUsers } from './index';
import { integrationConfig, withRecording } from '../../../test/config';
import { fetchAccountDetails } from '../account';

describe('#fetchUsers', () => {
  test(
    'should collect data',
    withRecording('fetchUsers', __dirname, async () => {
      const context = createMockStepExecutionContext<IntegrationConfig>({
        instanceConfig: integrationConfig,
      });

      await fetchAccountDetails(context);
      await fetchUsers(context);

      expect({
        numCollectedEntities: context.jobState.collectedEntities.length,
        numCollectedRelationships:
          context.jobState.collectedRelationships.length,
        collectedEntities: context.jobState.collectedEntities,
        collectedRelationships: context.jobState.collectedRelationships,
        encounteredTypes: context.jobState.encounteredTypes,
      }).toMatchSnapshot();

      const users = context.jobState.collectedEntities.filter((e) =>
        e._class.includes('User'),
      );

      expect(users.length).toBeGreaterThan(0);
      expect(users).toMatchGraphObjectSchema({
        _class: ['User'],
        schema: {
          additionalProperties: false,
          properties: {
            _type: { const: 'aquasec_user' },
            _rawData: {
              type: 'array',
              items: { type: 'object' },
            },
            id: { type: 'string' },
            email: { type: 'string' },
            confirmed: { type: 'boolean' },
            passwordReset: { type: 'boolean' },
            sendAnnouncements: { type: 'boolean' },
            sendScanResults: { type: 'boolean' },
            sendNewPlugins: { type: 'boolean' },
            sendNewRisks: { type: 'boolean' },
            admin: { type: 'boolean' },
            created: { type: 'string' },
            multiaccount: { type: 'boolean' },
          },
        },
      });
    }),
  );
});
