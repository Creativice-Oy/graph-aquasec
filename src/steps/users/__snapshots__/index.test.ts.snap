// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchUsers should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "aquasec_account:10630",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "additional_credits": 0,
            "associate_tags": false,
            "aws_marketplace_id": "",
            "block_default_group": false,
            "byok_bucket": null,
            "created": "2021-10-09T00:36:24.000Z",
            "csp_enabled": true,
            "csp_metadata": null,
            "current_plan": 57,
            "custom_cost": null,
            "disable_default_emails": false,
            "enforce_oauth": false,
            "global_suppressions_admin_only": false,
            "id": 10630,
            "integrations_admin_only": false,
            "just_in_time_domain": null,
            "just_in_time_domain_secondary": null,
            "login_timeout": null,
            "num_allowed_keys": 1,
            "presuppress_plugins": false,
            "renewal_date": null,
            "restrict_to_ips": null,
            "root_user": "austin.kelleher@jupiterone.dev",
            "source": "aquasec",
            "stripe_id": null,
            "suppressions_admin_only": false,
            "trial_end": "2021-10-23T00:36:24.000Z",
            "users_admin_only": false,
          },
        },
      ],
      "_type": "aquasec_account",
      "additionalCredits": 0,
      "associateTags": false,
      "awsMarketplaceId": "",
      "blockDefaultGroup": false,
      "createdOn": 1633739784000,
      "cspEnabled": true,
      "currentPlan": 57,
      "disableDefaultEmails": false,
      "displayName": "10630",
      "enforceOauth": false,
      "globalSuppressionsAdminOnly": false,
      "id": "10630",
      "integrationsAdminOnly": false,
      "manager": "austin.kelleher@jupiterone.dev",
      "name": "10630",
      "numAllowedKeys": 1,
      "presuppressPlugins": false,
      "rootUser": "austin.kelleher@jupiterone.dev",
      "source": "aquasec",
      "stripeId": undefined,
      "suppressionsAdminOnly": false,
      "trialEnd": 1634949384000,
      "usersAdminOnly": false,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "aquasec_group:12022",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created": "2021-10-09T00:36:24.000Z",
            "id": 12022,
            "name": "Default",
          },
        },
      ],
      "_type": "aquasec_group",
      "createdOn": 1633739784000,
      "displayName": "Default",
      "id": "12022",
      "name": "Default",
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "aquasec_group:12028",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created": "2021-10-09T15:56:46.000Z",
            "id": 12028,
            "name": "j1-test",
          },
        },
      ],
      "_type": "aquasec_group",
      "createdOn": 1633795006000,
      "displayName": "j1-test",
      "id": "12028",
      "name": "j1-test",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "aquasec_user:22971",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "account_admin": true,
            "confirmed": true,
            "created": "2021-10-09T00:36:25.000Z",
            "csp_roles": null,
            "dashboard": null,
            "email": "austin.kelleher@jupiterone.dev",
            "groups": Array [
              Object {
                "created": "2021-10-09T00:36:24.000Z",
                "id": 12022,
                "name": "Default",
              },
              Object {
                "created": "2021-10-09T15:56:46.000Z",
                "id": 12028,
                "name": "j1-test",
              },
            ],
            "id": 22971,
            "multiaccount": false,
            "password_reset": false,
            "provider": null,
            "send_announcements": false,
            "send_new_plugins": false,
            "send_new_risks": true,
            "send_scan_results": true,
          },
        },
      ],
      "_type": "aquasec_user",
      "active": true,
      "admin": true,
      "confirmed": true,
      "createdOn": 1633739785000,
      "displayName": "austin.kelleher@jupiterone.dev",
      "email": "austin.kelleher@jupiterone.dev",
      "id": "22971",
      "multiaccount": false,
      "name": "austin.kelleher@jupiterone.dev",
      "passwordReset": false,
      "sendAnnouncements": false,
      "sendNewPlugins": false,
      "sendNewRisks": true,
      "sendScanResults": true,
      "username": "austin.kelleher@jupiterone.dev",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "aquasec_user:22977",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "account_admin": false,
            "confirmed": true,
            "created": "2021-10-09T15:59:43.000Z",
            "csp_roles": null,
            "dashboard": null,
            "email": "austin.kelleher+j1test@jupiterone.dev",
            "groups": Array [
              Object {
                "created": "2021-10-09T00:36:24.000Z",
                "id": 12022,
                "name": "Default",
              },
              Object {
                "created": "2021-10-09T15:56:46.000Z",
                "id": 12028,
                "name": "j1-test",
              },
            ],
            "id": 22977,
            "multiaccount": false,
            "password_reset": false,
            "provider": null,
            "send_announcements": false,
            "send_new_plugins": false,
            "send_new_risks": true,
            "send_scan_results": true,
          },
        },
      ],
      "_type": "aquasec_user",
      "active": true,
      "admin": false,
      "confirmed": true,
      "createdOn": 1633795183000,
      "displayName": "austin.kelleher+j1test@jupiterone.dev",
      "email": "austin.kelleher+j1test@jupiterone.dev",
      "id": "22977",
      "multiaccount": false,
      "name": "austin.kelleher+j1test@jupiterone.dev",
      "passwordReset": false,
      "sendAnnouncements": false,
      "sendNewPlugins": false,
      "sendNewRisks": true,
      "sendScanResults": true,
      "username": "austin.kelleher+j1test@jupiterone.dev",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_account:10630",
      "_key": "aquasec_account:10630|has|aquasec_group:12022",
      "_toEntityKey": "aquasec_group:12022",
      "_type": "aquasec_account_has_group",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_account:10630",
      "_key": "aquasec_account:10630|has|aquasec_group:12028",
      "_toEntityKey": "aquasec_group:12028",
      "_type": "aquasec_account_has_group",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_account:10630",
      "_key": "aquasec_account:10630|has|aquasec_user:22971",
      "_toEntityKey": "aquasec_user:22971",
      "_type": "aquasec_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_account:10630",
      "_key": "aquasec_account:10630|has|aquasec_user:22977",
      "_toEntityKey": "aquasec_user:22977",
      "_type": "aquasec_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_group:12022",
      "_key": "aquasec_group:12022|has|aquasec_user:22971",
      "_toEntityKey": "aquasec_user:22971",
      "_type": "aquasec_group_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_group:12028",
      "_key": "aquasec_group:12028|has|aquasec_user:22971",
      "_toEntityKey": "aquasec_user:22971",
      "_type": "aquasec_group_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_group:12022",
      "_key": "aquasec_group:12022|has|aquasec_user:22977",
      "_toEntityKey": "aquasec_user:22977",
      "_type": "aquasec_group_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_group:12028",
      "_key": "aquasec_group:12028|has|aquasec_user:22977",
      "_toEntityKey": "aquasec_user:22977",
      "_type": "aquasec_group_has_user",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "aquasec_account",
    "aquasec_group",
    "aquasec_account_has_group",
    "aquasec_user",
    "aquasec_account_has_user",
    "aquasec_group_has_user",
  ],
  "numCollectedEntities": 5,
  "numCollectedRelationships": 8,
}
`;
